src/target/fhe/trans.rs

```diff
@@ -0,0 +1,362 @@
+
+                self.cache.insert(t.clone(), EmbeddedTerm::Bv);
+            }
+            _ => {
+                panic!("Non-field in embed_bv: {:?}", t);
+            }
+        }
+    }
+
+    // Given term `t`, type-check `t` is of type arr
+    fn check_arr(&self, t: &Term) {
+        match self
+            .cache
+            .get(t)
+            .unwrap_or_else(|| panic!("Missing expression for {:?}", t))
+        {
+            EmbeddedTerm::Arr => (),
+            _ => panic!("Check_arr failed"),
+        };
+    }
+
+    fn embed_arr(&mut self, t: Term) {
+        let s = self.term_to_stext_cnt.get(&t).unwrap();
+        match &t.op {
+            Op::Var(name, Sort::Array(_, _, len)) => {
+                // write to bytecode file
+                if !self.inputs.contains(&t) && self.md.input_vis.contains_key(name) {
+                    let term_name = ToFHE::get_var_name(&t);
+                    let enc = self.unwrap_vis(name);
+                    let op = "ARR_IN";
+
+                    let line = format!("3 1 {} {} {} {} {}\n", term_name, enc, len, s, op);
+                    self.bytecode_output.insert(0, line);
+                }
+                if !self.cache.contains_key(&t) {
+                    self.cache.insert(t.clone(), EmbeddedTerm::Arr);
+                }
+            }
+            Op::Const(Value::Array(arr)) => {
+                self.bytecode_output.push(format!("{} 1", arr.size));
+
+                for ival in arr.key_sort.elems_iter_values().take(arr.size) {
+                    let val = arr.select(&ival);
+                    if let Value::Array(_) = val.clone() {
+                        panic!("Const arr does not support multi-dim arrays")
+                    }
+                    self.bytecode_output.push(format!(" {}", val));
+                }
+
+                let op = "ARR_CONS";
+                let line = format!("{} {}\n", s, op);
+                self.bytecode_output.push(line);
+                self.cache.insert(t.clone(), EmbeddedTerm::Arr);
+            }
+            Op::Map(op) => {
+                self.bytecode_output.push(format!("{} 1", t.cs.len()));
+                for cterm in &t.cs {
+                    self.check_arr(cterm);
+                    self.bytecode_output
+                        .push(format!(" {}", self.term_to_stext_cnt.get(cterm).unwrap()));
+                }
+                let opstr = match **op {
+                    Op::BoolNaryOp(BoolNaryOp::Or) => "B_OR",
+                    Op::BoolNaryOp(BoolNaryOp::And) => "B_AND",
+                    Op::BoolNaryOp(BoolNaryOp::Xor) => "B_XOR",
+                    Op::BvNaryOp(BvNaryOp::Add) => "ADD",
+                    Op::BvNaryOp(BvNaryOp::Mul) => "MUL",
+                    _ => panic!("Map does not support operation {}", op),
+                };
+                self.bytecode_output.push(format!(" {} {}\n", s, opstr));
+                self.cache.insert(t.clone(), EmbeddedTerm::Arr);
+            }
```

Comment:
Is `Op::Store` is unsupported by this back-end?

Reply:
What about `Op::Select`? (although its implementation would not go in this function)

Reply:
The backend doesn't support Store and Select yet.

These operations are a bit complex because they require the rotation of the batched ciphertexts on the interpreter side. I am going to need some more time to implement these operations.